pars-fmt/src/lib.rs

```diff
@@ -0,0 +1,281 @@
+//! Parsing our format strings.
+//!
+//! Format strings are arbitrary strings containing 'field markers', indicating
+//! the location of fields. Field markers begin with `#{` and end with `}`,
+//! and contain the name of a struct field, or the index of a tuple field.
+//!
+//! Field markers must be separated by one or more dividing characters.
+//! During parsing, we scan the input looking for these dividing strings,
+//! which are discarded; the remaining subsections are associated with the
+//! matching fields, and parsed based on the field type.
+//!
+//! # Examples
+//!
+//!```ignore
+//! // A rect, represented as a space-delimited x, y, width, height
+//! // example input: "10 11 320 104"
+//!
+//! #[pars::fmt("#{x} #{y} #{width} #{height}")]
+//! struct Rect {
+//!     x: usize,
+//!     y: usize,
+//!     width: usize,
+//!     height: usize,
+//! }
+//!
+//!```
+
+use std::ops::Range;
+
+pub type Error = &'static str;
+
+enum State {
+    Ready,
+    InField,
+    /// member is the start position
+    InSeparator(usize),
+    Finished,
+}
+
+#[derive(Debug, Clone, PartialEq)]
+enum Token {
+    Field(String),
+    Separator(String),
+}
+
+struct Parser<'a> {
+    inp: &'a str,
```

Comment:
This name makes sense to someone who's used to thinking of pointers from C, but maybe we could call it `source` or somesuch?